processor/src/main/java/org/mapstruct/ap/internal/model/source/MappingOptions.java

```diff
@@ -54,13 +57,176 @@ public MappingOptions(Map<String, List<Mapping>> mappings, IterableMapping itera
     }
 
     /**
+     * creates empty mapping options
+     *
+     * @return empty mapping options
+     */
+    public static MappingOptions empty() {
+        return new MappingOptions( Collections.<String, List<Mapping>>emptyMap(), null, null, null,
+            Collections.<ValueMapping>emptyList() );
+    }
+
+    /**
+     * creates mapping options with only regular mappings
+     *
+     * @param mappings regular mappings to add
+     * @return MappingOptions with only regular mappings
+     */
+    public static MappingOptions forMappingsOnly( Map<String, List<Mapping>> mappings ) {
+        return new MappingOptions( mappings, null, null, null, Collections.<ValueMapping>emptyList() );
+
+    }
+
+    /**
      * @return the {@link Mapping}s configured for this method, keyed by target property name. Only for enum mapping
      * methods a target will be mapped by several sources. TODO. Remove the value list when 2.0
      */
     public Map<String, List<Mapping>> getMappings() {
         return mappings;
     }
 
+    /**
+     * The target references are popped. The MappingOptions are keyed on the unique first entries of the
+     * target references.
+     *
+     * So, take
+     *
+     * targetReference 1: propertyEntryX.propertyEntryX1.propertyEntryX1a
+     * targetReference 2: propertyEntryX.propertyEntryX2
+     * targetReference 3: propertyEntryY.propertyY1
+     * targetReference 4: propertyEntryZ
+     *
+     * will be popped and grouped into entries:
+     *
+     * propertyEntryX - MappingOptions ( targetReference1: propertyEntryX1.propertyEntryX1a,
+     *                                   targetReference2: propertyEntryX2 )
+     * propertyEntryY - MappingOptions ( targetReference1: propertyEntryY1 )
+     *
+     * The key will be the former top level property, the MappingOptions will contain the remainders.
+     *
+     * So, 2 cloned new MappingOptions with popped targetReferences. Also Note that the not nested targetReference4
+     * disappeared.
+     *
+     * @return See above
+     */
+    public Map<PropertyEntry, MappingOptions> groupByPoppedTargetRefences() {
+
+        // sort all mappings based on the top level name before popping
```

Comment:
I'll change it, but I think its correct. In English AFAIK: you sort stuff into groups.